xonsh/tools.py

```diff
@@ -92,6 +93,85 @@ def __init__(self, returncode, command, output=None, stderr=None,
         self.completed_command = completed_command
 
 
+def expandpath(path):
+    """
+    Performs environment variable / user expansion on a given path
+    if the relevant flag has been set.
+    """
+    env = getattr(builtins, '__xonsh_env__', os.environ)
+    if env.get('EXPAND_ENV_VARS', False):
+        # expand variables and use os.path.realpath to handle cases
+        # with relative paths like ../ or ./
+        path = os.path.expanduser(expandvars(path))
+    return os.path.realpath(path)
+
+
+def decode_bytes(path):
+    """
+    Tries to decode a path in bytes using XONSH_ENCODING if available,
+    otherwise using sys.getdefaultencoding().
+    """
+    env = getattr(builtins, '__xonsh_env__', os.environ)
+    enc = env.get('XONSH_ENCODING', sys.getdefaultencoding())
+    return path.decode(encoding=enc,
+                       errors=env.get('XONSH_ENCODING_ERRORS'))
+
+
+class EnvPath(MutableSequence):
+    """
+    A class that implements an environment path, which is a list of
+    strings. Provides a custom method that expands all paths if the
+    relevant env variable has been set.
+    """
+    def __init__(self, *args):
+        if len(args) > 0:
+            if isinstance(args[0], str):
+                self._l = args[0].split(os.pathsep)
+            elif isinstance(args[0], pathlib.Path):
+                self._l = [args[0]]
+            elif isinstance(args[0], bytes):
+                # decode bytes to a string and then split based on
+                # the default path separator
+                self._l = decode_bytes(args[0]).split(os.pathsep)
+            elif isinstance(args[0], Sequence):
+                if not all(isinstance(i, (str, bytes, pathlib.Path)) \
+                                      for i in args[0]):
+                    # make TypeError's message as informative as possible
+                    # when given an invalid initialization sequence
+                    raise TypeError(
+                            "EnvPath's initialization sequence should only "
+                            "contain str, bytes and pathlib.Path entries")
+                self._l = list(args[0])
+            else:
+                raise TypeError('EnvPath cannot be initialized with items '
+                                'of type %s' % type(args[0]))
```

Comment:
Whoops! You're right, I totally forgot about the case where `EnvPath` is called with more than one arguments. I suppose if a call like

```
EnvPath([path1, path2], [path3, path4])
```

occurs at somepoint this should produce an `EnvPath` containing `[path1, path2, path3, path4]`, right?